src-self-hosted/Module.zig

```diff
@@ -515,13 +528,47 @@ pub const Scope = struct {
         zir_module,
         /// .zig source code.
         file,
+        /// struct, enum or union, every .file contains one of these.
+        container,
         block,
         decl,
         gen_zir,
         local_val,
         local_ptr,
     };
 
+    pub const Container = struct {
+        pub const base_tag: Tag = .container;
+        base: Scope = Scope{ .tag = base_tag },
+
+        file_scope: *Scope.File,
+
+        /// Direct children of the file.
+        decls: ArrayListUnmanaged(*Decl),
+
+        // TODO implement container types and put this in a status union
+        // ty: Type
```

Comment:
One way we might end up doing this would be to have struct, enum, and union each have a `Scope.Container` field and so we would get access to the type (and other extra information via `@fieldParentPtr`). Not necessarily, but just throwing that idea out there.